File: photo-studio/app/report.py
"""Human-readable processing reports (per product and whole-batch)."""
from __future__ import annotations

from datetime import datetime, timezone

from app.models import BatchReport, ProcessingSettings, ProductProcessingReport

PRIVACY_NOTICE = (
    "Privacy notice: when Gemini processing is enabled, photographs are sent to Google's "
    "Gemini API for editing. Disable Gemini in the settings to keep all processing local."
)


def _timestamp() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")


def build_product_report(report: ProductProcessingReport, settings: ProcessingSettings) -> str:
    lines: list[str] = []
    lines.append(f"PROCESSING REPORT — {report.folder_name}")
    lines.append("=" * (len(lines[0])))
    lines.append(f"Generated: {_timestamp()}")
    lines.append(f"Mode: {settings.mode.value}")
    lines.append(f"Background style: {settings.background_style.value}")
    lines.append(f"Aspect ratio: {settings.aspect_ratio.value}")
    lines.append(
        f"Images: {len(report.results)} total | {report.success_count} enhanced | "
        f"{report.unsafe_count} kept original (safety) | {report.failure_count} failed"
    )
    lines.append("")

    for result in report.results:
        lines.append(f"- {result.filename}")
        status = "OK" if result.success else ("UNSAFE / ORIGINAL KEPT" if result.marked_unsafe else "FAILED")
        lines.append(f"    Status: {status}")
        if result.output_filename:
            lines.append(f"    Output: ENHANCED_PHOTOS/{result.output_filename}")
        if result.used_gemini:
            lines.append("    Enhanced by: Gemini AI")
        elif result.used_local_fallback:
            lines.append("    Enhanced by: Local processing (Pillow/OpenCV)")
        elif result.success:
            lines.append("    Enhanced by: none (resized only)")
        if result.similarity_score is not None:
            lines.append(f"    Preservation similarity score: {result.similarity_score:.2f}")
        for op in result.operations:
            lines.append(f"    - {op}")
        for warning in result.warnings:
            lines.append(f"    WARNING: {warning}")
        if result.error:
            lines.append(f"    ERROR: {result.error}")
        lines.append("")

    lines.append(PRIVACY_NOTICE if settings.use_gemini else "Gemini was disabled for this batch; all processing was local.")
    lines.append("")
    lines.append(
        "This report does not certify the garment's condition. Any defects visible in the "
        "original photographs (stains, holes, tears, wear, etc.) remain visible in the enhanced "
        "photographs by design."
    )
    return "\n".join(lines)


def build_batch_report(batch: BatchReport, settings: ProcessingSettings, extraction_warnings: list[str]) -> str:
    lines: list[str] = []
    title = "BATCH PROCESSING REPORT — Jordyn AI Photo Studio"
    lines.append(title)
    lines.append("=" * len(title))
    lines.append(f"Generated: {_timestamp()}")
    lines.append(f"Mode: {settings.mode.value}")
    lines.append(f"Gemini enabled: {settings.use_gemini}")
    lines.append(f"Local enhancement enabled: {settings.use_local_enhancement}")
    lines.append(f"Background style: {settings.background_style.value}")
    lines.append(f"Aspect ratio: {settings.aspect_ratio.value}")
    lines.append(f"Products processed: {len(batch.product_reports)}")
    lines.append(f"Total images: {batch.total_images}")
    lines.append(f"  Successfully enhanced: {batch.total_success}")
    lines.append(f"  Kept original (AI safety check failed): {batch.total_unsafe}")
    lines.append(f"  Failed: {batch.total_failure}")
    lines.append(f"Gemini API calls made: {batch.gemini_calls_made}")
    lines.append("")

    lines.append("Per-product summary:")
    for product_report in batch.product_reports:
        lines.append(
            f"  - {product_report.folder_name}: {len(product_report.results)} images, "
            f"{product_report.success_count} enhanced, {product_report.unsafe_count} kept original, "
            f"{product_report.failure_count} failed"
        )
    lines.append("")

    if extraction_warnings:
        lines.append("ZIP extraction warnings:")
        for w in extraction_warnings:
            lines.append(f"  - {w}")
        lines.append("")

    lines.append(PRIVACY_NOTICE if settings.use_gemini else "Gemini was disabled for this batch; all processing was local.")
    lines.append("")
    lines.append(
        "Strict preservation policy: this application never removes stains, holes, tears or other "
        "damage, never changes garment colour/shape/size, and never alters logos, labels or care "
        "tags. Any photo the AI could not confidently edit within these rules was left as the "
        "original image and flagged above."
    )
    return "\n".join(lines)

File: photo-studio-php/includes/report.php
<?php
declare(strict_types=1);

/** Human-readable processing reports (per product and whole-batch). */

const JPS_PRIVACY_NOTICE =
    "Privacy notice: when Gemini processing is enabled, photographs are sent to Google's " .
    'Gemini API for editing. Disable Gemini in the settings to keep all processing local.';

function jps_timestamp(): string
{
    return gmdate('Y-m-d H:i:s') . ' UTC';
}

function jps_product_success_count(array $productReport): int
{
    return count(array_filter($productReport['results'], fn($r) => $r['success'] && !$r['marked_unsafe']));
}

function jps_product_unsafe_count(array $productReport): int
{
    return count(array_filter($productReport['results'], fn($r) => $r['marked_unsafe']));
}

function jps_product_failure_count(array $productReport): int
{
    return count(array_filter($productReport['results'], fn($r) => !$r['success']));
}

function jps_batch_total_success(array $batchReport): int
{
    return array_sum(array_map('jps_product_success_count', $batchReport['product_reports']));
}

function jps_batch_total_unsafe(array $batchReport): int
{
    return array_sum(array_map('jps_product_unsafe_count', $batchReport['product_reports']));
}

function jps_batch_total_failure(array $batchReport): int
{
    return array_sum(array_map('jps_product_failure_count', $batchReport['product_reports']));
}

function jps_build_product_report(array $productReport, array $settings): string
{
    $lines = [];
    $title = "PROCESSING REPORT — {$productReport['folder_name']}";
    $lines[] = $title;
    $lines[] = str_repeat('=', strlen($title));
    $lines[] = 'Generated: ' . jps_timestamp();
    $lines[] = 'Mode: ' . $settings['mode'];
    $lines[] = 'Background style: ' . $settings['background_style'];
    $lines[] = 'Aspect ratio: ' . $settings['aspect_ratio'];
    $lines[] = sprintf(
        'Images: %d total | %d enhanced | %d kept original (safety) | %d failed',
        count($productReport['results']),
        jps_product_success_count($productReport),
        jps_product_unsafe_count($productReport),
        jps_product_failure_count($productReport)
    );
    $lines[] = '';

    foreach ($productReport['results'] as $result) {
        $lines[] = "- {$result['filename']}";
        $status = $result['success']
            ? 'OK'
            : ($result['marked_unsafe'] ? 'UNSAFE / ORIGINAL KEPT' : 'FAILED');
        $lines[] = "    Status: $status";
        if (!empty($result['output_filename'])) {
            $lines[] = "    Output: ENHANCED_PHOTOS/{$result['output_filename']}";
        }
        if ($result['used_gemini']) {
            $lines[] = '    Enhanced by: Gemini AI';
        } elseif ($result['used_local_fallback']) {
            $lines[] = '    Enhanced by: Local processing (GD)';
        } elseif ($result['success']) {
            $lines[] = '    Enhanced by: none (resized only)';
        }
        if ($result['similarity'] !== null) {
            $lines[] = sprintf('    Preservation similarity score: %.2f', $result['similarity']);
        }
        foreach ($result['operations'] as $op) {
            $lines[] = "    - $op";
        }
        foreach ($result['warnings'] as $warning) {
            $lines[] = "    WARNING: $warning";
        }
        if (!empty($result['error'])) {
            $lines[] = "    ERROR: {$result['error']}";
        }
        $lines[] = '';
    }

    $lines[] = $settings['use_gemini'] ? JPS_PRIVACY_NOTICE : 'Gemini was disabled for this batch; all processing was local.';
    $lines[] = '';
    $lines[] = 'This report does not certify the garment\'s condition. Any defects visible in the ' .
        'original photographs (stains, holes, tears, wear, etc.) remain visible in the enhanced ' .
        'photographs by design.';

    return implode("\n", $lines);
}

function jps_build_batch_report(array $batchReport, array $settings, array $extractionWarnings): string
{
    $lines = [];
    $title = 'BATCH PROCESSING REPORT — Jordyn AI Photo Studio (PHP edition)';
    $lines[] = $title;
    $lines[] = str_repeat('=', strlen($title));
    $lines[] = 'Generated: ' . jps_timestamp();
    $lines[] = 'Mode: ' . $settings['mode'];
    $lines[] = 'Gemini enabled: ' . ($settings['use_gemini'] ? 'true' : 'false');
    $lines[] = 'Local enhancement enabled: ' . ($settings['use_local_enhancement'] ? 'true' : 'false');
    $lines[] = 'Background style: ' . $settings['background_style'];
    $lines[] = 'Aspect ratio: ' . $settings['aspect_ratio'];
    $lines[] = 'Products processed: ' . count($batchReport['product_reports']);
    $lines[] = 'Total images: ' . $batchReport['total_images'];
    $lines[] = '  Successfully enhanced: ' . jps_batch_total_success($batchReport);
    $lines[] = '  Kept original (AI safety check failed): ' . jps_batch_total_unsafe($batchReport);
    $lines[] = '  Failed: ' . jps_batch_total_failure($batchReport);
    $lines[] = 'Gemini API calls made: ' . $batchReport['gemini_calls_made'];
    $lines[] = '';

    $lines[] = 'Per-product summary:';
    foreach ($batchReport['product_reports'] as $productReport) {
        $lines[] = sprintf(
            '  - %s: %d images, %d enhanced, %d kept original, %d failed',
            $productReport['folder_name'],
            count($productReport['results']),
            jps_product_success_count($productReport),
            jps_product_unsafe_count($productReport),
            jps_product_failure_count($productReport)
        );
    }
    $lines[] = '';

    if (!empty($extractionWarnings)) {
        $lines[] = 'ZIP extraction warnings:';
        foreach ($extractionWarnings as $w) {
            $lines[] = "  - $w";
        }
        $lines[] = '';
    }

    $lines[] = $settings['use_gemini'] ? JPS_PRIVACY_NOTICE : 'Gemini was disabled for this batch; all processing was local.';
    $lines[] = '';
    $lines[] = 'Strict preservation policy: this application never removes stains, holes, tears or ' .
        'other damage, never changes garment colour/shape/size, and never alters logos, labels or ' .
        'care tags. Any photo the AI could not confidently edit within these rules was left as the ' .
        'original image and flagged above.';

    return implode("\n", $lines);
}
